README.md
# Generic expression parser

This extension provides an API that can be used in other extensions for parsing expressions
with a syntax very much like the getText functionality of TypoScript. Please refer to the
documentation for more details.

ChangeLog
2017-06-06 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 2.1.0

2017-04-25 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Cleaned up code, verified compatibility with TYPO3 CMS 8

2016-03-14 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 2.0.0, resolves #75524

2016-03-14 Fabien Udriot  <fabien.udriot@ecodev.ch>

	* Added composer file, resolves #75053

2015-06-05 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Moved to namespaces, verified compatibility with TYPO3 CMS 7, resolves #67305
	* Moved documentation to reST, resolves #67314

2015-05-09 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Verified compatibility with TYPO3 CMS 6.2, resolves #58661
	* Released to TER as version 1.8.0, resolves #58662

2013-05-27 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Verified compatibility with TYPO3 6.1, resolves #48576
	* Released to TER as version 1.7.0

2013-03-26 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Corrected syntax error in function processing, resolves #46700
	* Verified compatibility with TYPO3 6.0, resolves #46698
	* Corrected erroneous examples in the manual, resolves #46701
	* Released to TER as version 1.6.0

2012-07-10 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 1.5.0

2012-06-26 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Cleaned up some code, tested compatibility with TYPO3 4.7, raised requirements to TYPO3 4.5, resolves #38389

2012-01-11 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Small code cleanup

2011-04-04 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added require statement to Fluid view helper, resolves #25740
	* Added support for inline notation, resolves #25742
	* Released to TER as version 1.4.0

2010-12-28 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 1.3.0

2010-10-20 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Add new function "quoteStr" (mapped to t3lib_db::quoteStr())

2010-09-28 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added proper handling of function processing when value is an array

2010-09-27 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Removed suggestion of extension "context"
	* Released to TER as version 1.2.3

2010-09-04 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Removed unnecessary usage of "context" extension hook

2010-08-25 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Corrected behavior of exceptions when using alternative expressions
	* Released to TER as version 1.2.2

2010-08-20 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Corrected abusive empty() check
	* Released to TER as version 1.2.1

2010-06-11 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added new function key "env"
	* Corrected a bug with multiple colons in an expression
	* Released to TER as version 1.2.0

2010-06-02 Fabien Udriot  <fabien.udriot@ecodev.ch>

	* Added View Helper for Fluid
	* Added shortcut key "fe_user"

2010-05-05 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Corrected parser behavior in case of empty string values
	* Released to TER as version 1.1.2

2010-02-02 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added new function "strftime"
	* Released to TER as version 1.1.1

2010-01-20 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added function calls on expressions
	* Fixed broken support for subexpressions
	* Released to TER as version 1.1.0

2010-01-20 Fabien Udriot  <fabien.udriot@ecodev.ch>

	* Added shortcut key "plugin"

2009-10-04 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 1.0.0


Classes/KeyProcessorInterface.php
<?php
namespace Cobweb\Expressions;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

/**
 * Interface which defines the method to implement when creating a special key processor for expressions.
 *
 * @author Francois Suter (Cobweb) <typo3@cobweb.ch>
 * @package TYPO3
 * @subpackage tx_expressions
 */
interface KeyProcessorInterface
{
    /**
     * This method must be implemented to handle the "indices" part it receives from the parser.
     *
     * Assuming an expression such as:
     *
     * mykey:foo|bar
     *
     * the method will receive the string foo|bar.
     * It is expected to return a value that makes sense or else throw an exception.
     *
     * @param string $indices The string to be interpreted
     * @return mixed The resulting value
     */
    public function getValue($indices);
}


Classes/Sample/FunctionProcessor.php
<?php
namespace Cobweb\Expressions\Sample;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

/**
 * Sample class providing an example user function for post-processing expressions.
 *
 * To activate that function, enter the following is some localconf file:
 *
 * $GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['expressions']['customFunction']['offset'] = 'Cobweb\\Expressions\\Sample\\FunctionProcessor->offset';
 *
 * To use it in an expression, enter something like:
 *
 * gp:foo|bar->offset:10
 *
 * This will take the foo[bar] GET/POST variable and offset it by 10
 *
 * @author Francois Suter (Cobweb) <typo3@cobweb.ch>
 * @package TYPO3
 * @subpackage tx_expressions
 */
class FunctionProcessor {
	/**
	 * Sample method to process expressions using the function-calling syntax
	 * i.e. this method might be called using:
	 *
	 * gp:foo|bar->offset:10
	 *
	 * The method receives an array of parameters and a reference to a dummy object
	 * (there's no real reference to a calling object, because the calling object is static).
	 * The array of parameters always contains the value from the expression first
	 * (i.e. in $parameters[0]).
	 * Any additional parameters required by the function come next. In this sample's
	 * case, the second parameter ($parameters[1]) is a number by which the value
	 * will be offset.
	 *
	 * The method *must* return the expression's value, even if unchanged.
	 *
	 * @param array	$parameters Indexed list of parameters. The first one is the expression's result.
	 * @param object $parentObject Dummy object. Not usable.
	 * @return mixed The processed value
	 */
	public function offset($parameters, $parentObject) {
		$value = (int)$parameters[0];
		$offset = 0;
		if (isset($parameters[1])) {
			$offset = (int)$parameters[1];
		}
		return $value + $offset;
	}
}


Classes/ValuePostProcessorInterface.php
<?php
namespace Cobweb\Expressions;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

/**
 * Interface which defines the method to implement when creating a hook to post-process parsed expressions.
 *
 * @author Francois Suter (Cobweb) <typo3@cobweb.ch>
 * @package TYPO3
 * @subpackage tx_expressions
 */
interface ValuePostProcessorInterface
{
    /**
     * This method must be implemented for post-processing a parsed value.
     *
     * It must return a value too, even if that is the unchanged input value.
     *
     * @param mixed $value The value that resulted from parsing an expression
     * @return mixed The resulting value
     */
    public function postprocessReturnValue($value);
}


Classes/ViewHelpers/EvaluateViewHelper.php
<?php
namespace Cobweb\Expressions\ViewHelpers;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

use Cobweb\Expressions\ExpressionParser;
use TYPO3\CMS\Fluid\Core\ViewHelper\AbstractViewHelper;

/**
 * This class is a Expressions view helper for the Fluid templating engine.
 *
 * In TypoScript templates there exists a function called “getText”, which makes it possible
 * to retrieve values from a great number of sources (GET/POST, TSFE, rootline, etc.).
 * Extension "expressions" provides a library which aims to reproduce this capability,
 * so that it can be used by developers where Typoscript is not used for whatever reasons.
 *
 * Like for the TypoScript function, an expression – at a minimum – is comprised of a key, a colon (:) and a value.
 * Example: tsfe:id
 *
 * Expression keys reference:
 *
 * config:    get a value from the “config” object, as stored in $TSFE->config['config']
 * date:      get values related to the current time, using formats from the PHP date() function.
 * extra:     see documentation
 * env:       get a value from the environment variables
 * fe_user:   get values from FE User
 * gp:        get a value from either the $_GET or $_POST superglobal arrays
 * page:      get a value related to the current page, as stored in $TSFE->page
 * plugin:    get a TypoScript property for a plugin as stored in $GLOBALS['TSFE']->tmpl->setup['plugin.'].
 * session:   get values from some structure stored in the temporary session
 * strtotime: get a timestamp by interpreting a human-readable date.
 * tsfe:      get a value from the $TSFE global variable
 * vars:      get a value from "internal" variables.
 *
 * More documentation can be found at
 * https://docs.typo3.org/typo3cms/extensions/expressions/2.0.0/User/ExpressionKeys/Index.html
 *
 * = Examples =
 *
 * Assuming that namespace has been declared as follows:
 * {namespace expression = Cobweb\Expressions\ViewHelpers}
 *
 * <code title="Simple string with expression to evaluate">
 * <expression:evaluate>Current page id is \{tsfe:id\}</expression:evaluate>
 * </code>
 * <output>
 * Current page id is 1
 * (assuming the current page has uid = 1, of course)
 * </output>
 * Note the escaped curly braces, so that Fluid is not confused
 *
 * <code title="Inline notation">
 * {expression:evaluate(expression:'Current user is \{fe_user:username\}')}
 * </code>
 * <output>
 * Current user is zaphod
 * (assuming the current FE user's username is "zaphod")
 * </output>
 *
 * @package TYPO3
 * @subpackage tx_expressions
 * @author Fabien Udriot <fabien.udriot@ecodev.ch>
 * @author Francois Suter <typo3@cobweb.ch>
 */
class EvaluateViewHelper extends AbstractViewHelper
{

    /**
     * Evaluates expression throughout the Expression Parser.
     *
     * @param string $expression Expression to be evaluated
     * @return string the evaluated string.
     */
    public function render($expression = null)
    {
        if ($expression === null) {
            $expression = $this->renderChildren();
        }
        // Replace escaped curly braces
        $searches[] = '\{';
        $replaces[] = '{';
        $searches[] = '\}';
        $replaces[] = '}';
        $expression = str_replace($searches, $replaces, $expression);
        // Evaluate and return
        return ExpressionParser::evaluateString($expression);
    }

}


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "expressions".
 *
 * Auto generated 06-06-2017 11:55
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array (
  'title' => 'Generic expression parser',
  'description' => 'A library for bringing into extensions something like the TypoScript getText function.',
  'category' => 'misc',
  'author' => 'Francois Suter (Cobweb)',
  'author_email' => 'typo3@cobweb.ch',
  'state' => 'stable',
  'uploadfolder' => 0,
  'createDirs' => '',
  'clearCacheOnLoad' => 0,
  'author_company' => '',
  'version' => '2.1.0',
  'constraints' => 
  array (
    'depends' => 
    array (
      'typo3' => '7.0.0-8.99.99',
    ),
    'conflicts' => 
    array (
    ),
    'suggests' => 
    array (
    ),
  ),
  '_md5_values_when_last_written' => 'a:31:{s:9:"ChangeLog";s:4:"4ae9";s:11:"LICENSE.txt";s:4:"6404";s:9:"README.md";s:4:"9cd9";s:13:"composer.json";s:4:"8988";s:12:"ext_icon.png";s:4:"cdfa";s:28:"Classes/ExpressionParser.php";s:4:"7929";s:33:"Classes/KeyProcessorInterface.php";s:4:"cf67";s:39:"Classes/ValuePostProcessorInterface.php";s:4:"b1d5";s:31:"Classes/Exception/Exception.php";s:4:"1e6f";s:36:"Classes/Sample/FunctionProcessor.php";s:4:"11c3";s:42:"Classes/ViewHelpers/EvaluateViewHelper.php";s:4:"f6e4";s:26:"Documentation/Includes.txt";s:4:"c83c";s:23:"Documentation/Index.rst";s:4:"93b4";s:26:"Documentation/Settings.yml";s:4:"0c08";s:33:"Documentation/Developer/Index.rst";s:4:"dbdc";s:37:"Documentation/Developer/Api/Index.rst";s:4:"2b57";s:44:"Documentation/Developer/Exceptions/Index.rst";s:4:"c6b4";s:39:"Documentation/Developer/Hooks/Index.rst";s:4:"61f7";s:46:"Documentation/Developer/NewFunctions/Index.rst";s:4:"4929";s:41:"Documentation/Developer/NewKeys/Index.rst";s:4:"7f4e";s:42:"Documentation/Developer/Security/Index.rst";s:4:"f277";s:43:"Documentation/Developer/Variables/Index.rst";s:4:"30f5";s:44:"Documentation/Developer/ViewHelper/Index.rst";s:4:"5965";s:36:"Documentation/Installation/Index.rst";s:4:"dd3c";s:36:"Documentation/Introduction/Index.rst";s:4:"67bc";s:37:"Documentation/KnownProblems/Index.rst";s:4:"056c";s:28:"Documentation/User/Index.rst";s:4:"73fc";s:43:"Documentation/User/ExpressionKeys/Index.rst";s:4:"74eb";s:45:"Documentation/User/ExpressionSyntax/Index.rst";s:4:"4ff9";s:41:"Documentation/User/FunctionKeys/Index.rst";s:4:"15d8";s:35:"Tests/Unit/ExpressionParserTest.php";s:4:"e0c6";}',
);

